[jax2tf] add rudimentary scatter support for enable_xla=False

oliverdutton opened the pull request:
I think this solves #9659.

Just checking now. It should work for simple indices where you index in only 1 dimension with your update. I've added quite a few guarding asserts. It still needs unit tests to be sure.

Checking #9659 test case now.

oliverdutton commented:
Minor bug in my asserts, fixing now

oliverdutton commented:
Asserts added back in, with tf.assert when tensors involved.

josephrocca commented:
Out of curiosity, I tried to use this fix to get the derivative of Flax CLIP exporting correctly, and before you added your asserts back in it failed on the unique index assertion:
```
AssertionError: Indices must be unique for this implementation. For other cases look at using something with tf.math.unsorted_segment_*
```
But with your asserts added back in I get this error:
```
    854     tf.assert_equal(tf.rank(operand), tf.rank(updates)+1, message='Updates must be 1 dimension less than operand')
    855     tf.assert_rank(scatter_indices, 1, message='Scatter indices must be 1 dimensional')
--> 856     assert (len(wd)==len(sd)==1) & (wd==sd) & (ud==tuple(range(max(ud)+1))), 'Unsupported scatter op'
    857     tf.assert_rank(operand, len(ud)+1, message='Unsupported scatter op')
    858     axis = wd[0] # axis the `scatter_indices` index into within `operand`

ValueError: max() arg is an empty sequence
```
So it looks like an extra assert might be needed above that line? Here's a minimal notebook: https://colab.research.google.com/drive/1SPeT3bBBSzZpAz0kqgfzvvYCsTXIPJIb

(It would be really cool if this PR added support for CLIP too, in case it's not too hard to support non-unique indices, but you've already been absurdly helpful so please ignore this if you don't have the bandwidth right now to do this!)

gnecula reviewed (commented):
Thank you very much for contributing this. It is one of the most complex cases.

Can you please add some tests?

You would do this in [primitives_harness.py](https://github.com/google/jax/blob/8e4cf253dc047d23cbbcb34a57a99fb82427c943/jax/experimental/jax2tf/tests/primitive_harness.py#L1229), add a `enable_xla` parameter to the harnesses for scatter. Without that parameter set, it defaults to `True`. We want to create a harness with both True/False.

A good example to follow is [`gather`](https://github.com/google/jax/blob/8e4cf253dc047d23cbbcb34a57a99fb82427c943/jax/experimental/jax2tf/tests/primitive_harness.py#L1103). Note that not all `gather` tests work with `enable_xla=False`, see how the `needs_xla` variable is used. I assume something similar can be done to the existing tests for scatter.

To run the tests, I do something like
```
pytest --durations=10 -r a --verbosity 1 -s jax/experimental/jax2tf/tests/primitives_test.py -k "scatter"
```

oliverdutton commented on jax/experimental/jax2tf/impl_no_xla.py:
```diff
@@ -792,10 +792,91 @@ def _dynamic_update_slice(operand, update, *start_indices,
+    wd = dimension_numbers.inserted_window_dims
+    sd = dimension_numbers.scatter_dims_to_operand_dims
+    ud = dimension_numbers.update_window_dims
+    tf.assert_equal(tf.rank(operand), tf.rank(updates)+1, message='Updates must be 1 dimension less than operand')
```
Shifted in latest commits

oliverdutton commented on jax/experimental/jax2tf/impl_no_xla.py:
```diff
@@ -851,10 +855,14 @@ def sparse_scatter(
+      raise error(f"Scatter does not support operands of type {dtype}")
+    if not ((scatter_indices.shape==(1,)) & (len(operand.shape) == len(updates.shape)+1) & (len(updates.shape) == len(ud)) &
+      (len(wd) == len(sd) == 1) & (wd == sd) & (ud == tuple(range(max(ud)+1)))):
+      raise error('Scatters of more than one row in one dimension are unsupported')
```
I think as long as the `lax.scatter_*` is being used correctly the first check `(scatter_indices.shape==(1,)` is all that's needed. The other inputs will only be wrong if someone is calling via lax.scatter_* directly rather than the .at[].* or numpy syntax which is rare and discouraged.

Should I make it just the first bit, or keep all these checks which ask if the scatter is valid in general.

gnecula replied:
I would keep all the checks, this makes it more locally-obvious that the code is correct

oliverdutton commented on jax/experimental/jax2tf/tests/primitive_harness.py:
```diff
@@ -1315,6 +1317,25 @@ def _make_scatter_harness(name,
 
+# Validate no XLA scatters
+for dtype in set(jtu.dtypes.all) - set(jtu.dtypes.complex) - set(jtu.dtypes.boolean):
+  for f_lax in [
```
I would add `lax.scatter` here but I get
```python
E       AssertionError: Harness group 'scatter' must be either part of 'harness_groups_no_limitations' or must have a custom Jax2TfLimitation.classmethod defined (see module docstring)
```
when I try to. 

Next commit fixes this.

oliverdutton commented:
`scatter` was overlooked in the list `scatter_add,...`. I've added it in and expanded the tests over it.

gnecula reviewed (approved):
Looks good, thanks!

josephrocca commented:
What should this code do if `dimension_numbers.update_window_dims` is an empty sequence? I'm still seeing the `ValueError: max() arg is an empty sequence` error [in this notebook](https://colab.research.google.com/drive/1SPeT3bBBSzZpAz0kqgfzvvYCsTXIPJIb) which tries to port CLIP with jax2tf. If it's not supported yet, then a more informative error could be thrown in this case?

oliverdutton commented:
Ah yes, I’d forgotten. I’ll add that in before it’s merged.

marcvanzee commented on jax/experimental/jax2tf/impl_no_xla.py:
```diff
@@ -792,10 +792,99 @@ def _dynamic_update_slice(operand, update, *start_indices,
-tf_impl_no_xla[lax.scatter_mul_p] = _unimplemented("scatter mul")
-tf_impl_no_xla[lax.scatter_add_p] = _unimplemented("scatter add")
+def get_move_to_leading_axis_bijector(axis):
+  '''
```
Nit: we usually use """ rather than '''.

marcvanzee commented on jax/experimental/jax2tf/impl_no_xla.py:
```diff
@@ -792,10 +792,99 @@ def _dynamic_update_slice(operand, update, *start_indices,
-tf_impl_no_xla[lax.scatter_max_p] = _unimplemented("scatter max")
-tf_impl_no_xla[lax.scatter_mul_p] = _unimplemented("scatter mul")
-tf_impl_no_xla[lax.scatter_add_p] = _unimplemented("scatter add")
+def get_move_to_leading_axis_bijector(axis):
```
I am a bit confused by this name and the docstring. Given that it is only used once, perhaps it is easier to understand if you simply define the `shift` function and then define `fwd = shift` and `bwd = partial(shift, inverse=True)` below where you need it?

marcvanzee commented on jax/experimental/jax2tf/impl_no_xla.py:
```diff
@@ -792,10 +792,99 @@ def _dynamic_update_slice(operand, update, *start_indices,
-tf_impl_no_xla[lax.scatter_add_p] = _unimplemented("scatter add")
+def get_move_to_leading_axis_bijector(axis):
+  '''
+  Bijector taking an array x, of rank `operand_rank`.
+  Forward shift the axis `axis` to the leading axis
+
+  so for array `operand_rank`=4 and `axis=4`. The last axis is
+  moved for the front: permutation - [3,0,1,2]
```
This docstring is a bit confusing, because this function actually returns two functions, which is not really clear. Also the name `get_move_to_leading_axis_bijector` is maybe not very enlightening.

One option is to just define a function `shift_axis_forward(operation, axis, inverse=False)` and define `_fwd = partial(shift_axis, axis=axis)` and `_bwd = partial(_forward, inverse=True)` when you need it below (you only use them once anyway).

oliverdutton replied:
Agreed, it's a terrible name. Will alter.

marcvanzee commented on jax/experimental/jax2tf/impl_no_xla.py:
```diff
@@ -792,10 +792,99 @@ def _dynamic_update_slice(operand, update, *start_indices,
+    '''
+    suboperand = tf.gather_nd(operand, scatter_indices)
+    updated_suboperand = update_op(suboperand, updates)
+    return tf.tensor_scatter_nd_update(operand, scatter_indices[None], updated_suboperand[None])
```
Perhaps it is useful to explain in a comment why you need to add an empty dimension to `scatter_indices` and `updated_suboperand`.

oliverdutton replied:
Altered doc string. It comes down to:
XLAGather has dimension numbers which deal with specifying batch dimensions and which dims the scatters go in to. tf.tensor_scatter_nd_update does not have dimension numbers so has to infer everything from the shapes, it indexs from the front only and forcibly has at least one batch dimension.

For example, If you have 4D array, but are indexing into the last 2 dims e.g. x[:,:,0,0] you need to shift the last two dims to the front to use a tf equivalent, you can't just 'pass' on a dim.

marcvanzee commented on jax/experimental/jax2tf/impl_no_xla.py:
```diff
@@ -792,10 +792,99 @@ def _dynamic_update_slice(operand, update, *start_indices,
+    _in_avals: Sequence[core.ShapedArray],
+    _out_aval: core.ShapedArray):
+    '''
+    This wraps around the scatter function
+    This works for functions which index only in one of the axes, it shifts the axis to
+    be operated on to the lead so a lead specific implementation can be used
+    '''
```
```suggestion
    Wrapper around the scatter function.
    This works for functions that index in only one of the axes, it shifts the axis to
    be operated on to the front so a front-specific implementation can be used.
    '''
```

oliverdutton replied:
Syntax improvements added

marcvanzee reviewed (changes requested):
Looks really great, thanks for adding this! Added a few small comments.

oliverdutton commented:
My dim asserts are actually a bit off now I have time to think about it, I'll fix those

oliverdutton commented:
Expanded support up to  .at[:, ((3,4),(5,5)), 3] style scatters

oliverdutton commented:
I expanded support to non-unique indices where the index is to depth 1. i.e `x.at[:, ((0,1,1),)].`. I've added tests for a number of complex cases. These all pass.

I ran @josephrocca's [FlaxCLIPModel example](https://colab.research.google.com/drive/1SPeT3bBBSzZpAz0kqgfzvvYCsTXIPJIb#scrollTo=BU6ySvOGFmFt) and got the wrong answer, I got zeros everywhere. Pure jax gets happy gradients everywhere. It contains 6 scatter-add's I can see in the jax_pr. Possibly one of them has failed but I have no visibility on it failing, it does not error. It may be nothing to do with the scatters, but it seems likely.


```python
fft:f32[1] = scatter-add[
      dimension_numbers=ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
      indices_are_sorted=True
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      unique_indices=True
      update_consts=()
      update_jaxpr={ lambda ; ffu:f32[] ffv:f32[]. let
          ffw:f32[] = add ffu ffv
        in (ffw,) }
    ] ffs ffr 1.0
    fga:f32[1,1] = scatter-add[
      dimension_numbers=ScatterDimensionNumbers(update_window_dims=(0,), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
      indices_are_sorted=True
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      unique_indices=True
      update_consts=()
      update_jaxpr={ lambda ; ffu:f32[] ffv:f32[]. let
          ffw:f32[] = add ffu ffv
        in (ffw,) }
    ] ffz ffl ffy
    fge:f32[1,1,1] = scatter-add[
      dimension_numbers=ScatterDimensionNumbers(update_window_dims=(0, 1), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
      indices_are_sorted=True
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      unique_indices=True
      update_consts=()
      update_jaxpr={ lambda ; ffu:f32[] ffv:f32[]. let
          ffw:f32[] = add ffu ffv
        in (ffw,) }
    ] fgd fff fgc
    fic:f32[1,50,768] = scatter-add[
      dimension_numbers=ScatterDimensionNumbers(update_window_dims=(0, 1), inserted_window_dims=(1,), scatter_dims_to_operand_dims=(1,))
      indices_are_sorted=True
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      unique_indices=True
      update_consts=()
      update_jaxpr={ lambda ; ffu:f32[] ffv:f32[]. let
          ffw:f32[] = add ffu ffv
        in (ffw,) }
    ] fib dbe fia
          hcs:f32[3,224,40] = scatter-add[
            dimension_numbers=ScatterDimensionNumbers(update_window_dims=(0, 1), inserted_window_dims=(2,), scatter_dims_to_operand_dims=(2,))
            indices_are_sorted=False
            mode=GatherScatterMode.PROMISE_IN_BOUNDS
            unique_indices=False
            update_consts=()
            update_jaxpr={ lambda ; ffu:f32[] ffv:f32[]. let
                ffw:f32[] = add ffu ffv
              in (ffw,) }
          ] hcr hco hcq
          hcv:f32[3,40,40] = scatter-add[
            dimension_numbers=ScatterDimensionNumbers(update_window_dims=(0, 2), inserted_window_dims=(1,), scatter_dims_to_operand_dims=(1,))
            indices_are_sorted=False
            mode=GatherScatterMode.PROMISE_IN_BOUNDS
            unique_indices=False
            update_consts=()
            update_jaxpr={ lambda ; ffu:f32[] ffv:f32[]. let
                ffw:f32[] = add ffu ffv
              in (ffw,) }
          ] hcu hcn hct
```

oliverdutton commented:
I ran through the jax_pr's extracted the scatter-add's and added tests. None fail. Looks like the problem in that FlaxCLIP model gradients going to zero in jax2tf.convert aren't down to the scatter ops. 

I am done with adding functionality to this pull request. Please advise.

gnecula commented:
Thank you @oliverdutton. Can you please squash the many commits into one, or a small number of meaningful ones?

For the error that remains, can either you or @josephrocca update #9659 (which we'll keep open)?

oliverdutton committed:
tests: add lax.scatter to tests

oliverdutton committed:
feat: support subset of scatters with no XLA

Supports:
- Scatters with unique indices e.g..at[:, ((3,4),(5,5)), 3].add
- Contiguous scatters with non-unique indices of single depth e.g..at[:, ((2,2,2),)].add
